File: android/src/main/java/com/scandit/datacapture/flutter/id/ScanditFlutterDataCaptureIdProxyPlugin.java
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2024- Scandit AG. All rights reserved.
 */
package com.scandit.datacapture.flutter.id;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import com.scandit.datacapture.flutter.core.BaseFlutterPlugin;
import com.scandit.datacapture.flutter.core.utils.FlutterEmitter;
import com.scandit.datacapture.frameworks.core.FrameworkModule;
import com.scandit.datacapture.frameworks.core.locator.ServiceLocator;
import com.scandit.datacapture.frameworks.id.IdCaptureModule;
import com.scandit.datacapture.frameworks.id.listeners.FrameworksIdCaptureListener;

import java.util.concurrent.atomic.AtomicInteger;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodChannel;

public class ScanditFlutterDataCaptureIdProxyPlugin extends BaseFlutterPlugin implements FlutterPlugin, ActivityAware {
    private final static FlutterEmitter idEmitter = new FlutterEmitter(IdCaptureMethodHandler.EVENT_CHANNEL_NAME);

    private static final AtomicInteger activePluginInstances = new AtomicInteger(0);

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        activePluginInstances.incrementAndGet();
        super.onAttachedToEngine(binding);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        activePluginInstances.decrementAndGet();
        super.onDetachedFromEngine(binding);
    }

    @Override
    protected int getActivePluginInstanceCount() {
        return activePluginInstances.get();
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        setupEventChannels();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        disposeEventChannels();
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        onAttachedToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        disposeEventChannels();
    }

    @Override
    protected void setupMethodChannels(FlutterPluginBinding binding, ServiceLocator<FrameworkModule> serviceLocator) {
        MethodChannel channel = createChannel(
                binding,
                IdCaptureMethodHandler.METHOD_CHANNEL_NAME
        );
        channel.setMethodCallHandler(new IdCaptureMethodHandler(serviceLocator));
        registerChannel(channel);
    }

    @Override
    protected void setupModules(FlutterPlugin.FlutterPluginBinding binding) {
        IdCaptureModule idCaptureModule = resolveModule(IdCaptureModule.class);
        if (idCaptureModule != null) return;

        idCaptureModule = IdCaptureModule.create(
                FrameworksIdCaptureListener.create(idEmitter)
        );
        idCaptureModule.onCreate(binding.getApplicationContext());

        registerModule(idCaptureModule);
    }

    private void setupEventChannels() {
        FlutterPluginBinding binding = getCurrentBinding();
        if (binding != null) {
            idEmitter.addChannel(binding.getBinaryMessenger());
        }
    }

    private void disposeEventChannels() {
        FlutterPluginBinding binding = getCurrentBinding();
        if (binding != null) {
            idEmitter.removeChannel(binding.getBinaryMessenger());
        }
    }

    @VisibleForTesting
    public static void resetActiveInstances() {
        activePluginInstances.set(0);
    }
}

File: android/src/main/kotlin/com/scandit/datacapture/flutter/id/ScanditFlutterDataCaptureIdProxyPlugin.kt
package com.scandit.datacapture.flutter.id

import com.scandit.datacapture.flutter.core.extensions.getMethodChannel
import com.scandit.datacapture.flutter.core.utils.FlutterEmitter
import com.scandit.datacapture.frameworks.core.locator.DefaultServiceLocator
import com.scandit.datacapture.frameworks.id.IdCaptureModule
import com.scandit.datacapture.frameworks.id.listeners.FrameworksIdCaptureListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import java.lang.ref.WeakReference
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/** ScanditFlutterDataCaptureIdProxyPlugin */
class ScanditFlutterDataCaptureIdProxyPlugin : FlutterPlugin, ActivityAware {
    companion object {
        @JvmStatic
        private val lock = ReentrantLock()

        @JvmStatic
        private var isPluginAttached = false
    }

    private val serviceLocator = DefaultServiceLocator.getInstance()

    private var idCaptureMethodChannel: MethodChannel? = null

    private var flutterPluginBinding: WeakReference<FlutterPluginBinding?> = WeakReference(null)

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        flutterPluginBinding = WeakReference(binding)
        onAttached()
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        flutterPluginBinding = WeakReference(null)
        onDetached()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        onAttached()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetached()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttached()
    }

    override fun onDetachedFromActivity() {
        onDetached()
    }

    private fun onAttached() {
        lock.withLock {
            if (isPluginAttached) {
                disposeModule()
            }

            val flutterBinding = flutterPluginBinding.get() ?: return

            setupModule(flutterBinding)

            isPluginAttached = true
        }
    }

    private fun onDetached() {
        lock.withLock {
            disposeModule()
            isPluginAttached = false
        }
    }

    private fun setupModule(binding: FlutterPluginBinding) {
        val eventEmitter = FlutterEmitter(
            EventChannel(
                binding.binaryMessenger,
                IdCaptureMethodHandler.EVENT_CHANNEL_NAME
            )
        )
        val idCaptureModule = IdCaptureModule(
            FrameworksIdCaptureListener(eventEmitter)
        ).also { module ->
            module.onCreate(binding.applicationContext)

            idCaptureMethodChannel = binding.getMethodChannel(
                IdCaptureMethodHandler.METHOD_CHANNEL_NAME
            ).also {
                it.setMethodCallHandler(IdCaptureMethodHandler(module))
            }
        }
        serviceLocator.register(idCaptureModule)
    }

    private fun disposeModule() {
        serviceLocator.remove(IdCaptureModule::class.java.name)?.onDestroy()
        idCaptureMethodChannel?.setMethodCallHandler(null)
    }
}
